### PR TITLE
[SMST] feat: Use compact SMST proofs

### PR DESCRIPTION
**Pull request by red-0ne**
## Summary

This PR changes the proofs being sent in `MsgSubmitPRoof` and stored by the `SubmitProof` keeper from `SparseMerkleClosestProof` to `SparseCompactMerkleClosestProof`.

## Issue

`Proofs` one of the most block-space consuming primitives. The SMST support compact proofs that could help reduce their sizes.

- #758 

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [x] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code

### CODE REVIEW

**Review by Olshansk — APPROVED** [review]
Amazed that everyone just worked!

Any chance you can add a log or a unit test to show the difference in size?

I'm REALLy curious and can be done in a followup.

**Comment by red-0ne** [comment]
> Any chance you can add a log or a unit test to show the difference in size?

@Olshansk, I'm really grateful that you called this out.

I added this test [/pkg/relayer/session/sessiontree_test.go](https://github.com/pokt-network/poktroll/pull/823/files#diff-d13257bc0fb6f968fa7ca4f63a57cb899a302c9a5f22d63f87a67f3f708014c2)

Which shows that there is barely any improvement in the proof size :cry: 

I added a (gzip) compressed proofs to the test to have something to compare with.

Running for different leaf sizes, 1000 run each and accumulating size, this a sample result

```
numLeaf=10: compactionRatio: 0.990917, compressionRatio: 0.804937
numLeaf=100: compactionRatio: 0.989091, compressionRatio: 0.803551
numLeaf=1000: compactionRatio: 0.992390, compressionRatio: 0.805643
numLeaf=10000: compactionRatio: 0.989311, compressionRatio: 0.803930
numLeaf=100000: compactionRatio: 0.992038, compressionRatio: 0.805471
numLeaf=1000000: compactionRatio: 0.990228, compressionRatio: 0.804995
```

File: pkg/relayer/session/sessiontree_test.go [hunk 1]
```diff
@@ -1,3 +1,96 @@
...
+	"crypto/rand"
+	"testing"
+
+	"github.com/pokt-network/poktroll/pkg/crypto/protocol"
+	"github.com/pokt-network/smt"
+	"github.com/pokt-network/smt/kvstore/pebble"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:15]: #PUC to link to the results: https://github.com/pokt-network/poktroll/pull/823#issuecomment-2363987920

File: pkg/relayer/session/sessiontree_test.go [hunk 2]
```diff
@@ -1,3 +1,96 @@
...
+	// Run the test for different number of leaves.
+	for numLeaf := 10; numLeaf <= 1000000; numLeaf *= 10 {
+		// We run the test 1000 times for each number of leaves to remove the randomness bias.
+		cumulativeProofSize := 0
+		cumulativeCompactProofSize := 0
+		cumulativeGzippedProofSize := 0
+		for numLeaf := 0; numLeaf <= 1000; numLeaf++ {
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:22]: We're using the same variable `numLeaf` internally and externally. Please fix.

File: pkg/relayer/session/sessiontree_test.go [hunk 3]
```diff
@@ -1,3 +1,96 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
+	// Run the test for different number of leaves.
+	for numLeaf := 10; numLeaf <= 1000000; numLeaf *= 10 {
+		// We run the test 1000 times for each number of leaves to remove the randomness bias.
+		cumulativeProofSize := 0
+		cumulativeCompactProofSize := 0
+		cumulativeGzippedProofSize := 0
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:21]:
```suggestion
		cumulativeProofSize := 0
		cumulativeCompactProofSize := 0
		cumulativeGzippedProofSize := 0
		// We run the test 1000 times for each number of leaves to remove the randomness bias.
```

File: pkg/relayer/session/sessiontree_test.go [hunk 4]
```diff
@@ -1,3 +1,96 @@
...
+
+	"github.com/pokt-network/poktroll/pkg/crypto/protocol"
+	"github.com/pokt-network/smt"
+	"github.com/pokt-network/smt/kvstore/pebble"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
+	// Run the test for different number of leaves.
+	for numLeaf := 10; numLeaf <= 1000000; numLeaf *= 10 {
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:17]:
```suggestion
	for numLeafs := 10; numLeafs <= 1000000; numLeafs *= 10 {
```

File: pkg/relayer/session/sessiontree_test.go (hunk 2 above)
Comment (Olshansk) [reply pkg/relayer/session/sessiontree_test.go:22]: Fix the naming for booth loop variables

File: pkg/relayer/session/sessiontree_test.go [hunk 5]
```diff
@@ -1,3 +1,96 @@
...
+			require.NoError(t, err)
+
+			// Create the proof.
+			proof, err := trie.ProveClosest(path)
+			require.NoError(t, err)
+
+			proofBz, err := proof.Marshal()
+			require.NoError(t, err)
+
+			// Accumulate the proof size over 1000 runs.
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:50]: Don't put numbers like this in the comments.

If you (or another developer) ever update the loop, there's no chance you'll update the comment.

Comments should be "ever green".

If they're not "ever green", they should involve a TODO.

File: pkg/relayer/session/sessiontree_test.go [hunk 6]
```diff
@@ -1,3 +1,96 @@
...
+			cumulativeProofSize += len(proofBz)
+
+			// Generate the compacted proof.
+			compactProof, err := smt.CompactClosestProof(proof, &trie.TrieSpec)
+			require.NoError(t, err)
+
+			compactProofBz, err := compactProof.Marshal()
+			require.NoError(t, err)
+
+			// Accumulate the compact proof size over 1000 runs.
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:60]: Ditto

File: pkg/relayer/session/sessiontree_test.go [hunk 7]
```diff
@@ -1,3 +1,96 @@
...
+			gzipWriter := gzip.NewWriter(&buf)
+			_, err = gzipWriter.Write(proofBz)
+			require.NoError(t, err)
+			err = gzipWriter.Close()
+			require.NoError(t, err)
+
+			// Accumulate the gzipped proof size over 1000 runs.
+			cumulativeGzippedProofSize += len(buf.Bytes())
+
+			//t.Logf(
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:74]: Either delete or merge in the things that we need

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Comment by Olshansk** [comment]
@red-0ne Thanks for putting in the extra test. This is fine and what I expected.

Want to use this as a learning lesson. Nothing we don't already know but something I want to reiterate:

1. `Compact proof` sounds sexy, eh? This is a classic example of "sounds sexy to an engineer" but not actually useful.
     - Probabilistic proof is what will actually ensure scalability
2. Understanding what's happening under the hood
     - We inherited "Compact Proofs" from the library we forked
     - I didn't implement it, neither did Harry, neither did you
     - None of us looked in the source code
     - I'm 100% sure we can make it more efficient, but again, it "sounds sexy", right?
3. The importance of visibility / observability
    - This doesn't always need to be a dashboard
    - It doesn't always need to be telemetry
    - It could be a log or a unit test
    - This is a great example of using `t.Log`
    - This raises the importance of a dashboard for probabilistic proofs

File: pkg/relayer/session/sessiontree_test.go [hunk 8]
```diff
@@ -12,23 +12,34 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+const (
+	maxLeafs      = 10000
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:16]: #PUC details for both of these constants

File: x/proof/keeper/proof_validation_test.go [hunk 9]
```diff
@@ -611,8 +611,10 @@ func TestEnsureValidProof_Error(t *testing.T) {
 				)
 
 				// Extract relayHash to check below that it's difficulty is insufficient
-				sparseMerkleClosestProof := &smt.SparseMerkleClosestProof{}
-				err = sparseMerkleClosestProof.Unmarshal(proof.ClosestMerkleProof)
+				err = sparseCompactMerkleClosestProof.Unmarshal(proof.ClosestMerkleProof)
+				require.NoError(t, err)
+				var sparseMerkleClosestProof *smt.SparseMerkleClosestProof
+				sparseMerkleClosestProof, err = smt.DecompactClosestProof(sparseCompactMerkleClosestProof, &protocol.SmtSpec)
 				require.NoError(t, err)
```
Comment (Olshansk) [inline x/proof/keeper/proof_validation_test.go:618]:
```suggestion

				var sparseMerkleClosestProof *smt.SparseMerkleClosestProof
				sparseMerkleClosestProof, err = smt.DecompactClosestProof(sparseCompactMerkleClosestProof, &protocol.SmtSpec)
				require.NoError(t, err)

```

File: pkg/relayer/session/sessiontree_test.go [hunk 10]
```diff
@@ -68,29 +79,19 @@ func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
...
-			// guaranteed to always reduce the proof size.
-			//require.Less(t, len(compactProofBz), len(proofBz))
 		}
 
-		//t.Logf(
-		//	"numLeaf=%d: cumulativeProofSize: %d, cumulativeCompactProofSize: %d, cumulativeGzippedProofSize: %d",
-		//	numLeaf, cumulativeProofSize, cumulativeCompactProofSize, cumulativeGzippedProofSize,
-		//)
+		compactionRatio := float32(cumulativeCompactProofSize) / float32(cumulativeProofSize)
+		compressionRatio := float32(cumulativeGzippedProofSize) / float32(cumulativeProofSize)
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:87]:
```suggestion
		// Calculate how much more efficient compact SMT proofs are
		compactionRatio := float32(cumulativeCompactProofSize) / float32(cumulativeProofSize)

		// Calculate how much more efficient gzipped SMT proofs are
		compressionRatio := float32(cumulativeGzippedProofSize) / float32(cumulativeProofSize)
```

File: pkg/relayer/session/sessiontree_test.go [hunk 11]
```diff
@@ -68,29 +79,19 @@ func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
...
+
+		// Gzip compression is more efficient than compaction.
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:89]:
```suggestion
		// Gzip compression is more efficient than SMT compaction.
```

File: pkg/relayer/session/sessiontree_test.go [hunk 12]
```diff
@@ -68,29 +79,19 @@ func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
...
+		require.Greater(t, compactionRatio, compressionRatio)
```
Comment (Olshansk) [inline pkg/relayer/session/sessiontree_test.go:90]: This seems to contradict the comment above, no?

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/relayer/session/sessiontree_test.go (hunk 12 above)
Comment (red-0ne) [reply pkg/relayer/session/sessiontree_test.go:90]: Inverted how the ratio is calculated, now the higher the better.

**Review by Olshansk — APPROVED** [review]

**Comment by h5law** [comment]
> @red-0ne Thanks for putting in the extra test. This is fine and what I expected.
> 
> Want to use this as a learning lesson. Nothing we don't already know but something I want to reiterate:
> 
> 1. `Compact proof` sounds sexy, eh? This is a classic example of "sounds sexy to an engineer" but not actually useful.
>    
>    * Probabilistic proof is what will actually ensure scalability
> 2. Understanding what's happening under the hood
>    
>    * We inherited "Compact Proofs" from the library we forked
>    * I didn't implement it, neither did Harry, neither did you
>    * None of us looked in the source code
>    * I'm 100% sure we can make it more efficient, but again, it "sounds sexy", right?
> 3. The importance of visibility / observability
>    
>    * This doesn't always need to be a dashboard
>    * It doesn't always need to be telemetry
>    * It could be a log or a unit test
>    * This is a great example of using `t.Log`
>    * This raises the importance of a dashboard for probabilistic proofs

Little late to this but just going through my notifications. I actually did implement the CompactClosestProof type to fit the pattern with the other proof types. Had some brief ideas about how they could be compacted properly (was in a rush when implementing this):

1. Clip the `Path` field to `Depth` bits
2. Replace the `FlippedBits` byte matrix with a bit mask where `1`s represent bits flipped, the `[][]byte` can be replaced with `uint{Depth}` as each bit represents an index (up to 4 `uint64`s for a full 256-bit worst case encoding)
3. Instead of encoding the `Depth` as the `[]byte` form of the original `int` use a single `byte` (this assumes a 256 bit hasher or figure out a 9-bit type to generalise and include 512-bit hashers)

Realisticallly this method is super optimisable, I didn't really try to optimise it as it was always don't pre-optimise. I'm sure you guys will think of even more, actually doing compression on the encoded bits would be a interesting idea once they are encoded better.

**Pull request merged**